node/lib/DeviceAccessor.ts

```diff
@@ -0,0 +1,241 @@
+
+import { DeviceInterface } from "./DeviceInterface";
+import { IDevice } from "./ITranslator";
+
+import { EventEmitter } from "events";
+
+/**
+ * Provides reflection-style access to device properties and methods via device interfaces.
+ */
+export class DeviceAccessor {
+
+    /**
+     * Gets the value of a property on a device, using a specified interface.
+     *
+     * @param {IDevice} device  Device instance obtained from a translator
+     * @param {(string | DeviceInterface)} interfaceName  Name of the interface used
+     *     to access the device, or a DeviceInterface object
+     * @param {string} propertyName  Name of the property to get
+     * @returns {Promise<any>} Value returned by the property getter; the value is
+     *     expected to conform to the JSON schema for the property as specified
+     *     by the interface
+     */
+    public static async getPropertyAsync(
+            device: IDevice,
+            interfaceName: string | DeviceInterface,
+            propertyName: string): Promise<any> {
+        let deviceInterface = DeviceAccessor.getDeviceInterface(device, interfaceName);
+        DeviceAccessor.validateMemberName(propertyName);
+
+        let value: any = deviceInterface[propertyName];
+        if (typeof value === "undefined") {
+            let methodName = "get" + DeviceAccessor.capitalize(propertyName);
+            let getPropertyMethod: any = deviceInterface[methodName];
+            if (typeof getPropertyMethod === "function") {
+                value = getPropertyMethod.call(deviceInterface);
```

Comment:
It's subtle, and maybe not intuitive due to the way we are essentially doing reflection in JavaScript. `getPropertyMethod` is a JavaScript Function object, so it could be possible to call `getPropertyMethod()` directly. The problem with that is the `this` parameter would not be passed in that case. The `Function.call()` method allows setting the value of this to `deviceInterface` when invoking the Function.

---
In reply to: [71902043](https://github.com/opent2t/opent2t/pull/1#discussion_r71902043) [](ancestors = 71902043)